ui_test/src/diff.rs

```diff
@@ -0,0 +1,134 @@
+            }
+            let skipped = self.skipped_lines.len() - CONTEXT * 2;
+            match skipped {
+                0 => {}
+                1 => eprintln!(" {}", self.skipped_lines[CONTEXT]),
+                _ => eprintln!("... {skipped} lines skipped ..."),
+            }
+            for line in self.skipped_lines.iter().rev().take(CONTEXT).rev() {
+                eprintln!(" {line}");
+            }
+        }
+        self.skipped_lines.clear();
+    }
+
+    fn skip(&mut self, line: &'a str) {
+        self.skipped_lines.push(line);
+    }
+
+    fn print_prev(&mut self) {
+        if let Some(l) = self.prev_left.take() {
+            self.print_left(l);
+        }
+    }
+
+    fn print_left(&self, l: &str) {
+        eprintln!("{}{}", "-".red(), l.red());
+    }
+
+    fn print_right(&self, r: &str) {
+        eprintln!("{}{}", "+".green(), r.green());
+    }
+
+    fn row(&mut self, row: Result<&'a str>) {
+        match row {
+            Left(l) => {
+                self.print_skip();
+                self.print_prev();
+                self.prev_left = Some(l);
+            }
+            Both(l, _) => {
+                self.print_prev();
+                self.skip(l);
+            }
+            Right(r) => {
+                if let Some(l) = self.prev_left.take() {
+                    // If the lines only add chars or only remove chars, display an inline diff
+                    let diff = chars(l, r);
+                    let mut seen_l = false;
+                    let mut seen_r = false;
+                    for char in &diff {
+                        match char {
+                            Left(l) if !l.is_whitespace() => seen_l = true,
+                            Right(r) if !r.is_whitespace() => seen_r = true,
+                            _ => {}
+                        }
+                    }
+                    if seen_l && seen_r {
+                        // the line both adds and removes chars, print both lines, but highlight their differences
+                        eprint!("{}", "-".red());
+                        for char in &diff {
+                            match char {
+                                Left(l) => eprint!("{}", l.to_string().red()),
+                                Right(_) => {}
+                                Both(l, _) => eprint!("{}", l),
+                            }
+                        }
+                        eprintln!();
+                        eprint!("{}", "+".green());
+                        for char in &diff {
+                            match char {
+                                Left(_) => {}
+                                Right(r) => eprint!("{}", r.to_string().green()),
+                                Both(l, _) => eprint!("{}", l),
+                            }
+                        }
+                        eprintln!();
+                    } else {
+                        eprint!("{}", "~".yellow());
```

Comment:
Please add a comment explaining this. I have never seen a tilde in a diff...

Reply:
Oh is this the "inline diff" case? The comments make it sound like `if let Some(l) = self.prev_left.take()` is the inline diff case, but then the "display both lines" is *within* that branch so the comment seems misleading.